### ISSUE TITLE
Calculate gradient 

### ISSUE THREAD

**Issue by Esme-lxy** [lwwu2/diver#4]
Hi!
First of all, thank you very much for your fantastic work and code!  
I am trying to optimize the c2w matrix(rotation and translation {R, t}). Now after loss.backward(), the gradient of R can be calculated, R can be optimized, but the gradient of t is calculated as None. Do you have any suggestions for this？

**Comment by lwwu2** [comment]
Hi!
Thank you for your interests in our work.
The 't' is passed as the input of the ray-voxel intersection, which is implemented as a cuda kernel that does not has back propagation function. I think that is the reason why there is no gradient for the translation. 

A quick but not elegant fix could be taking the ray-voxel intersection $x$ then calculating the corresponding ray depth to the ray origin (translation) $o$ as $t=|o-x|_2$ and re-calculate ray-voxel intersection with gradient $x=o+dt$. This might introduce small error to the final intersection location but should be fine. The code would be like (after line 107 of model/diver.py): 
```
def f(o,d,x):
    t = (x-o).norm(dim=-1).data
    x = o + d*t
    return x

ii = torch.where(mask)[0]

coord_in = coord_in*self.voxel_size + self.xyzmin
coord_in = f(os[ii],ds[ii],coord_in)
coord_in = (coord_in-self.xyzmin)/self.voxel_size

coord_out = coord_out*self.voxel_size + self.xyzmin
coord_out= f(os[ii],ds[ii],coord_out)
coord_out= (coord_out-self.xyzmin)/self.voxel_size
```
Let us know if you have more questions.

**Comment by Esme-lxy** [comment]
Thanks for your reply very much!
